starfish/image/_filter/zero_by_channel_magnitude.py

```diff
@@ -0,0 +1,111 @@
+        e.g., DARTFISH, SEQFISH, etc., this filter allows for the explicit zeroing
+        out of pixels, for each round, where there is insufficient signal magnitude across channels.
+
+
+        Parameters
+        ----------
+        thresh : int
+            pixels in each round that have a L2 norm across channels below this threshold
+            are set to 0
+
+        normalize : bool
+            if True, this scales all rounds to have unit L2 norm across channels
+        """
+        self.thresh = thresh
+        self.normalize = normalize
+
+    @classmethod
+    def add_arguments(cls, group_parser: argparse.ArgumentParser) -> None:
+        group_parser.add_argument(
+            '--thresh', type=int, help='minimum magnitude threshold for pixels across channels')
+
+    def run(self, stack: ImageStack, in_place: bool = True, verbose=False) -> Optional[ImageStack]:
+        """Perform filtering of an image stack
+
+        Parameters
+        ----------
+        stack : ImageStack
+            Stack to be filtered.
+        in_place : bool
+            if True, process ImageStack in-place, otherwise return a new stack
+        verbose : bool
+            if True, report on the percentage completed during processing (default = False)
+
+        Returns
+        -------
+        Optional[ImageStack] :
+            if in-place is False, return the results of filter as a new stack
+
+        """
+
+        channels_per_round = stack._data.groupby(Indices.ROUND.value)
+        channels_per_round = tqdm(channels_per_round) if verbose else channels_per_round
+
+        if not in_place:
+            new_stack = deepcopy(stack)
+            return self.run(new_stack, in_place=True)
+
+        # compute channel magnitude mask
+        for r, dat in channels_per_round:
+            # nervous about how xarray orders dimensions so i put this here explicitly ....
+            dat = dat.transpose(Indices.CH.value,
+                                Indices.Z.value,
+                                Coordinates.Y.value,
+                                Coordinates.X.value
+                                )
+            # ... to account for this line taking the norm across axis 0, or the channel axis
+            ch_magnitude = np.linalg.norm(dat, ord=2, axis=0)
+            magnitude_mask = ch_magnitude >= self.thresh
+
+            # apply mask and optionally, normalize by channel magnitude
+            for c in range(stack.num_chs):
+                ind = {Indices.ROUND: r, Indices.CH: c}
+                stack._data[ind] = stack._data[ind] * magnitude_mask
+
+                if self.normalize:
+                    stack._data[ind] = np.divide(stack._data[ind],
+                                                 ch_magnitude,
+                                                 where=magnitude_mask
+                                                 )
+
+        return stack
+
+# TODO @ajc the code below fulfills the same task without xarray.
```

Comment:
@ambrosejcarr i can't tell if i prefer the numpy array implementation or the xarray implementation? which one do you prefer? which one do you think is faster?

Reply:
I find the xarray code about 1000x easier to follow. 

Thankfully, it is also faster!

![image](https://user-images.githubusercontent.com/5239152/44243147-8d741d00-a19b-11e8-92d8-1daca014ff5b.png)

Reply:
this surprises me. i thought my old numpy code would for sure be faster. any idea why groupby is faster? i also think it's WAY easier to read which is awesome.

Reply:
I figured numpy would be faster as well -- my guess is that your code may be  copying or allocating arrays that xarray has found a way to avoid. 

Given that it's slower, I stopped thinking about it 😄

Reply:
lol ok i'll nuke the comment.